extract/spotify_api.py

```diff
@@ -0,0 +1,239 @@
+        tracks.append(track)
+    return tracks
+
+
+def get_artist_followers(artist_id: str, headers: dict) ->tuple:
+    """Takes an artist id and gets the popularity rating and follower count for that artist"""
+    result = get(f"{SPOTIFY_BASE_URL}artists/{artist_id}", headers=headers, timeout=10)
+    result = json.loads(result.content)
+    popularity = result["popularity"]
+    follower_count = result["followers"]["total"]
+    genres = result["genres"]
+    return (popularity, follower_count, genres)
+
+
+def get_track_popularity(track_id: str, headers: dict) -> tuple:
+    """Takes a track id and gets the popularity rating of that track"""
+    result = get(f"{SPOTIFY_BASE_URL}tracks/{track_id}", headers=headers, timeout=10)
+    result = json.loads(result.content)
+    if "popularity" in result:
+        popularity = result["popularity"]
+    else:
+        popularity = None
+    return popularity
+
+
+def get_track_audio_features(track_id: str, headers: dict) -> tuple:
+    """Takes a track id and gets danceability, energy, valence, tempo, and speechiness scores"""
+    result = get(f"{SPOTIFY_BASE_URL}audio-features/{track_id}", headers=headers, timeout=10)
+    result = json.loads(result.content)
+    danceability = result["danceability"]
+    energy = result["energy"]
+    valence = result["valence"]
+    tempo = result["tempo"]
+    speechiness = result["speechiness"]
+    return (danceability, energy, valence, tempo, speechiness)
+
+
+def get_db_connection():
+    """Connects to the database"""
+    try:
+        conn = psycopg2.connect(
+            user = os.getenv("DB_USER"),
+            password = os.getenv("DB_PASSWORD"),
+            host = os.getenv("DB_HOST"),
+            port = os.getenv("DB_PORT"),
+            database = os.getenv("DB")
+            )
+        print("Connected")
+        return conn
+    except Exception as e:
+        print(e)
+        print("Error connecting to database.")
+
+
+def add_track_data(data: list[dict]) -> list[dict]:
+    """Takes in data on tracks and inserts track details into the track table"""
+    for track in data:
+        with conn, conn.cursor(cursor_factory=RealDictCursor) as cur:
+            sql_input = "INSERT INTO track (track_name, track_danceability, track_energy, \
+                track_valence, track_tempo, track_speechiness, in_spotify, in_tiktok, \
+                    tiktok_rank, spotify_rank, track_spotify_id)\
+                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s) ON CONFLICT DO NOTHING"
+            vals = [track["name"], track["danceability"], track["energy"], track["valence"], \
+                    track["tempo"], track["speechiness"], track["in_spotify"], \
+                    track["in_tiktok"], track["tiktok_rank"], \
+                    track["spotify_rank"], track["id"]]
+            cur.execute(sql_input, vals)
+            conn.commit()
+        add_track_popularity(track["id"], track["popularity"])
```

Comment:
Instead of this for loop many we could implement an execute_values()?

`with conn, conn.cursor(cursor_factory=RealDictCursor) as cur:
         sql_input = "INSERT INTO track (track_name, track_danceability, track_energy, \
                track_valence, track_tempo, track_speechiness, in_spotify, in_tiktok, \
                    tiktok_rank, spotify_rank, track_spotify_id)\
                VALUES (%s) ON CONFLICT DO NOTHING"`
           execute_values(cur, sql_input, data)
            conn.commit`
But currently the code works fine so not top priority just something to look into if we have a bit of spare time